### PR TITLE
Docs: Fix Source block snippet updates

### LINKED ISSUE #22679: [Bug]: Boolean control is not persistent
### Describe the bug

As the title suggests, boolean controls are not persistent when the page is reloaded. On the one hand, it logs unnecessarily many warnings and on the other hand, direct links to stories with control values in it cannot be sent to someone else.

See the Reproduction section for more information.

### To Reproduce

https://stackblitz.com/edit/github-fbee7n?file=src%2Fstories%2FButton.stories.js

**Preparation:** Do not use the inline preview but open the preview in a new tab (button on the upper right). Only in this way it is possible for you to see the changes in the URL.

**Successful case:**
1. Go to the `Button/Primary` story
2. Clear the logs in the browser dev tools
3. Click on the `primary` control to change it to `false`.
    - Notice that `&args=primary:!false` is added to the URL
    - Notice this warning in the logs:
      ```
      Omitted potentially unsafe URL args.

      More info: https://storybook.js.org/docs/react/writing-stories/args#setting-args-through-the-url
      ```
4. Reload the page
    - Notice the `&args=primary:!false` is still in the URL and the `primary` control is set to `false`
    - Notice that the warning from 3. above is logged again.

**Failing case:**
1. Go to the `Button/Secondary` story
2. Clear the logs in the browser dev tools
3. Click on the `primary` control to change it to `true`.
    - Notice that `&args=primary:!true` is added to the URL
    - Notice this warning in the logs:
      ```
      Omitted potentially unsafe URL args.

      More info: https://storybook.js.org/docs/react/writing-stories/args#setting-args-through-the-url
      ```
4. Reload the page
    - Notice the `&args=primary:!true` is no longer in the URL and the `primary` control is set to `false`
    - Notice that the warning from 3. above is logged again.

### System

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chakAs3**
Closes #22679 #22645

## What I did

Many users reported the same bug #22679.
in case the story has a boolean arg going back and forth between true/false. will not update the Source block.

previous code compares the newSource and current Source Object that contains all the sources using the stringified args as id,  so if you toggle between 2 hashes that are already in the sources, you will not get any difference. 

the fix is to compare the last source snippet with the new one 

## How to test

 Run any sandbox the bug related to SourceContainer  e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access Any  story

-->

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: scripts/package.json [hunk 1]
```diff
@@ -148,7 +148,7 @@
     "ts-node": "^10.9.1",
     "tsup": "^6.7.0",
     "type-fest": "^3.4.0",
-    "typescript": "5.0.4",
+    "typescript": "~4.9.3",
```
Comment (shilman) [inline scripts/package.json:151]: What's up with this version change? Should this be a separate PR?

**Review by shilman — CHANGES_REQUESTED** [review]
Hi @chakAs3 !! Thanks so much for fixing this -- this looks like a good one.

Can you please add a test for this that fails with the old code and succeeds with the new code? I believe there is a bug and that this fixes it, but TBH I don't understand the change, so I'd like to see it on a real example in CI.

File: scripts/package.json (hunk 1 above)
Comment (chakAs3) [reply scripts/package.json:151]: No i did not touch this actually,  i don't know how it got there, actually it does not matter the version of typescript on scripts project i guess,

**Comment by chakAs3** [comment]
> Hi @chakAs3 !! Thanks so much for fixing this -- this looks like a good one.
> 
> Can you please add a test for this that fails with the old code and succeeds with the new code? I believe there is a bug and that this fixes it, but TBH I don't understand the change, so I'd like to see it on a real example in CI.

Ok, I see, I will try the issue happening on the  Docs source snippet, i can do unit tests but a visual story, i need to figure out how to access the block displayed source. cause the model is changing and the event is fired, the only issue is the guard to optimise rendering the snippet even though i keep the optimisation just changing the condition, i don't really see why we have to optimise this, it is not an expensive operation at all

**Comment by shilman** [comment]
@chakAs3 I think you can add an e2e test here https://github.com/storybookjs/storybook/blob/fc3b4b7843b77707e9c8936bf0a7880bf344f1c8/code/e2e-tests/addon-docs.spec.ts#L12

**Comment by chakAs3** [comment]
> @chakAs3 I think you can add an e2e test here
> 
> https://github.com/storybookjs/storybook/blob/fc3b4b7843b77707e9c8936bf0a7880bf344f1c8/code/e2e-tests/addon-docs.spec.ts#L12

Yes I can add unit-test

**Comment by chakAs3** [comment]
hi @shilman  it is ready if you can review it. there a PR with e2e test to show the bug

**Comment by shilman** [comment]
Thanks so much for this @chakAs3. I've merged the changes from the other PR into this one to verify the fix. I've also targeted the other branch so it's now a "telescoping" PR.

**Comment by tmeasday** [comment]
Ok, now I understand the issue. I don't think this is the right fix.

The situation:

1. We have a `SourceContainer` - a context provider that needs to change the context (triggering a re-render of all context consumers) whenever it hears about a new source value.
2. We have the `Source` block, that is a consumer of the context that renders the source for the current args value of the given story.

The bug is that the only reactivity involved is from 1, the `Source` block does not currently re-render when the args change. So only we we see a new value of args we haven't seen before, does the container get a new source and triggers a re-render.

This PR fixes it by making the source container trigger re-renders (of *all* consumers) whenever it sees even sources it has seen before. Basically it undoes an optimization to achieve that. This seems a bit perverse. Really the `Source` block needs to take responsibility for re-rendering when the args change.

I'd suggest we do that by something similar to what the `useArgs` hook does: https://github.com/storybookjs/storybook/blob/30a91c1b5fc61805596171a2923c2c354370f020/code/ui/blocks/src/blocks/useArgs.ts#L14-L41

Concievably we could just call that hook and not even look at the result to get the right reactivity in the source block. I can confirm that this fixes the test case (although it breaks the rules of hooks, so it's a bit more complex than that unfortunately):

image

cc @JReinhold who might have a good solution in mind.

**Comment by chakAs3** [comment]
@tmeasday Ok I got it now, for me this is kind of caching the source snippets, so here is the real issue is rendering the right source. but should rerender anyway, I don't know how the Source block works, I will look into it

**Comment by chakAs3** [comment]
@tmeasday  to be honest I don't see why we need to optimise the render here. optimising to render is the role of the renderer, Here react anyway if you give the same source no changes react will not rerender ( react does diffing itself so no DOM rerender).
In fact, we should not decide after firing SNIPPET_RENDERED , if we rerender or not. if a source transformer package needs to rerender a snippet for any reason not necessarily  args changes it can be params or any option provided by the plugin,
the  SourceContainer or SourceBlock component should rerender.

**Comment by tmeasday** [comment]
> to be honest I don't see why we need to optimise the render here.

I'm not super concerned about optimizing, it's just that there's an entanglement between the `Source` block and the container that doesn't need to be there. It means the `Source` block depends on behaviour of the container that isn't necessarily "correct".

What that means is in the future, someone might come along and make a change to the container that makes sense in isolation (like optimizing, or something else?) and then breaks the `Source` block. It just seems like bad code hygiene for the block to depend on the container's reactivity to ensure its own works correctly.

> I may have plugin that switches between two modes using context.parameters ( is the case for vue source decorator now)

I'm not sure I understand. The parameter can't change without changing story, right?

**Comment by chakAs3** [comment]
> > to be honest I don't see why we need to optimise the render here.
> 
> I'm not super concerned about optimizing, it's just that there's an entanglement between the `Source` block and the container that doesn't need to be there. It means the `Source` block depends on behaviour of the container that isn't necessarily "correct".
> 
> What that means is in the future, someone might come along and make a change to the container that makes sense in isolation (like optimizing, or something else?) and then breaks the `Source` block. It just seems like bad code hygiene for the block to depend on the container's reactivity to ensure its own works correctly.
> 
> > I may have plugin that switches between two modes using context.parameters ( is the case for vue source decorator now)
> 
> I'm not sure I understand. The parameter can't change without changing story, right?

I mean it is possible that I don't change args, but I still need to rerender the snippet, I want my source to include script setup tag for example, nothing to do with args or rendered story

**Comment by chakAs3** [comment]
> What that means is in the future, someone might come along and make a change to the container that makes sense in isolation (like optimizing, or something else?) and then breaks the `Source` block. It just seems like bad code hygiene for the block to depend on the container's reactivity to ensure its own works correctly.

100% Exact Container should not decide how Source Block behaves, should provide state whenever got changed. here it is true all blocks will get notified but that ok even if they rerender React handle the render optimisation by diffing, to make sure that the changed block will rerender properly to Real DOM

**Comment by tmeasday** [comment]
> These two snippets are for the same exact story.

I don't get it. `withScript` is a parameter? You can't change parameters in a story -- so it can't be the same exact story.

**Comment by chakAs3** [comment]
> > These two snippets are for the same exact story.
> 
> I don't get it. `withScript` is a parameter? You can't change parameters in a story -- so it can't be the same exact story.

yes but withScript does not impact the story render, only the source decorator, it is ok if you considered it not the same story, but  Source Container use only the args as hashed id.

**Comment by tmeasday** [comment]
But what I am saying is if the parameter is different it will a different story, thus stored in the source container differently. 

In either case the source container will update if it gets a new value for a given story/args' source so I'm not seeing the relevance.

**Comment by chakAs3** [comment]
> In either case the source container will update if it gets a new value for a given story/args'

Yes that is what it should be and what I did, I removed the optimisation test and just call setSources to update, and Each block will figure out what to render

**Comment by shilman** [comment]
@chakAs3 Jumping in late here. Looking through the conversation it looks like this PR re-renders all source snippets on the page every time there is a `SNIPPET_RENDERED` event, whereas @tmeasday is concerned that this fix should be located in the Source block since the block can access everything it needs to make that determination but is currently unaware of when the args change.

I looked the code a bit and it's not clear the best way to circumvent the rules of hooks with [hacky workarounds](https://stackoverflow.com/questions/67827350/calling-a-react-hook-conditionally), so I'd say let's go with @chakAs3 's fix for now. The rules of hooks thing will go away in Storybook 8.0 when we get rid of the deprecated code, so I'm going to add a comment to update things then.

**Comment by chakAs3** [comment]
> @chakAs3 Jumping in late here. Looking through the conversation it looks like this PR re-renders all source snippets on the page every time there is a `SNIPPET_RENDERED` event, whereas @tmeasday is concerned that this fix should be located in the Source block since the block can access everything it needs to make that determination but is currently unaware of when the args change.
> 
> I looked the code a bit and it's not clear the best way to circumvent the rules of hooks with [hacky workarounds](https://stackoverflow.com/questions/67827350/calling-a-react-hook-conditionally), so I'd say let's go with @chakAs3 's fix for now. The rules of hooks thing will go away in Storybook 8.0 when we get rid of the deprecated code, so I'm going to add a comment to update things then.

Excited for Storybook 8.0.

For me, the block should be only responsible for the snippet display behavior ( layout, style , actions: show/hide. , copy ..)
the optimization. should be done in the Decorator level, which is the case now for react and vue at least 
we only fire SNIPPET_RENDERED if the args have changed 
```ts
export const sourceDecorator = (storyFn: any, context: StoryContext<Renderer>) => {
  const skip = skipSourceRender(context);
  const story = storyFn();

  watch(
    () => context.args,
    () => {
      if (!skip) {
        generateSource(context);
      }
    },
    { immediate: true, deep: true }
  );
  return story;
};
```
so I really watching the changes, but again as you said the hooks implementation may mess with all this, as we need to call them every time, it is a temporary fix, but the work. will be more on the other side before. firing the event maybe
